### DISCUSSION TITLE
How to store results of parse_many for later use?

### DISCUSSION THREAD

**Discussion by misos1** [simdjson/simdjson#1986] [General]
**Is your feature request related to a problem? Please describe.**

Seems after use of `parse_many` the result cannot be obtained through parser through `parser.doc` and moving parser will probably invalidate what is returned from `parse_many` (suppose we need to move it for example into promise). Is there a better way than to use for example `unique_ptr`?

**Comment by lemire** [comment]
Maybe there is a misunderstanding... after calling `parse_many`, you get an instance of ` simdjson::dom::document_stream` (or an error code). The ` simdjson::dom::document_stream` instance does not contain the parsed input.

I recommend you explain your concrete need.

**Comment by misos1** [comment]
I need to parse json and store it for later in a `promise`. Seems I cannot just store `parser` because seems there is no way to obtain result from it so I need to store both `parser` and `document_stream` and seems I cannot move `parser` because this invalidates `document_stream` so one solution that comes to mind is to have `parser` in `unique_ptr`.

**Comment by lemire** [comment]
Yes, you can store the parser in a `unique_ptr`.

**Reply by misos1** [reply]
And seems also source string must be stored along with `parser` and `document_stream`. If it is `std::string` then it needs to be in `unique_ptr`. So I was curious if there is any better way. When I use `parse` then I can just store parsers and retrieve documents through `parser.doc` or use `parse_into_document` and store just documents without need to wrap in unique_ptr in both cases.

**Comment by lemire** [comment]
Why don't you just delay the creation of the parser until you need it?

**Reply by misos1** [reply]
I overlooked that it parses lazily.

**Reply by lemire** [reply]
In the DOM model, once have a document, it is parsed, but if you have a document stream, then you only have access to one parsed document at a time (at most).
